Src/Plugins/Security/SEC101_043.NuGetCredentialsValidator.cs

```diff
@@ -0,0 +1,154 @@
+            {
+                byte[] byteArray = Encoding.ASCII.GetBytes($"{username}:{password}");
+                httpClientWithCredentials.DefaultRequestHeaders.Authorization =
+                    new AuthenticationHeaderValue("Basic", Convert.ToBase64String(byteArray));
+
+                foreach (string host in hosts)
+                {
+                    if (!Uri.IsWellFormedUriString(host, UriKind.Absolute))
+                    {
+                        continue;
+                    }
+
+                    using (HttpResponseMessage responseWithNoCredentials = httpClientNoCredentials
+                    .GetAsync(host, HttpCompletionOption.ResponseHeadersRead)
+                    .GetAwaiter()
+                    .GetResult())
+                    {
+                        switch (responseWithNoCredentials.StatusCode)
+                        {
+                            case HttpStatusCode.OK:
+                                // Credentials not needed, this method of verification is indeterminate.
+                                return ReturnUnknownAuthorization(ref message, username);
+                            case HttpStatusCode.Unauthorized:
+                            case HttpStatusCode.Forbidden:
+                                // Credentials may resolve this, try again with them.
+                                using (HttpResponseMessage responseWithCredentials = httpClientWithCredentials
+                                .GetAsync(host, HttpCompletionOption.ResponseHeadersRead)
+                                .GetAwaiter()
+                                .GetResult())
+                                {
+                                    switch (responseWithCredentials.StatusCode)
+                                    {
+                                        case HttpStatusCode.OK:
+                                            // Credentials resolved the forbidden/unauthorized message we received.
+                                            return ReturnAuthorizedAccess(ref message, username);
+                                        case HttpStatusCode.Forbidden:
+                                        case HttpStatusCode.Unauthorized:
+                                            // Credentials didn't fix the error, but it may be that we're missing
+                                            // something in the request.
+                                            return ReturnUnauthorizedAccess(ref message, username);
+                                        default:
+                                            break;
+                                    }
+                                }
+
+                                break;
+                            default:
+                                break;
+                        }
+                    }
+                }
+            }
+
+            return ValidationState.Unknown;
```

Comment:
>ValidationState.Unknown [](start = 19, length = 23)

we don't do this at function exit. instead, all handling s/be in the default cases in the switch statements such that this specific line of code is unreachable. please note that you should be returning the unexpected response code received for any code paths where we are, in fact, surprised by the response code. see other rules for examples of this.

Reply:
Refactored, so...done, kinda.

---
In reply to: [611239974](https://github.com/microsoft/sarif-pattern-matcher/pull/372#discussion_r611239974) [](ancestors = 611239974)